### ISSUE TITLE
[BUG] Improve texture selection for xcomposite

### ISSUE THREAD

**Issue by kkartaltepe** [obsproject/obs-studio#2937]
## Platform
Operating system and version: Linux
OBS Studio version: 25.0.8
Drivers: amdgpu-pro

## Expected Behavior
Texture format chosen in https://github.com/obsproject/obs-studio/blob/master/plugins/linux-capture/xcompcap-main.cpp#L423 should support `glCopyImageSubData` from the source window format.

## Current Behavior
When users use amdgpu-pro drivers the selected format is almost always incompatible. Resulting in
`glCopyImageSubData failed, glGetError returned GL_INVALID_OPERATION(0x502)`

## Steps to Reproduce
Install amdgpu-pro drivers.
Try to capture chromium.
Uninstall drivers and install mesa drivers.
See that capture works again.

## Additional information
We should probably log the actual visuals used by windows and what we select so we can cross-reference glxinfo during support.

**Comment by lfzawacki** [comment]
I'm hitting this bug as well. Is there some way to work around it before OBS lands a fix? I'd also like to keep the amdgpu-pro drivers.

**Comment by kkartaltepe** [comment]
The workaround is to uninstall the amdgpu-pro drivers.

**Comment by lfzawacki** [comment]
> The workaround is to uninstall the amdgpu-pro drivers.

Ok, my mistake was thinking that the amdgpu drivers were equal to amdgpu-**pro** drivers. I actually uninstalled them and used the 'amdgpu-install' file from the download provided from the AMD website and got it to work. Sorry for the noise.

**Comment by gustopn** [comment]
So. I just hit this bug too. CentOS8 AMDGPU-PRO and gives:
`error: glCopyImageSubData failed, glGetError returned GL_INVALID_OPERATION(0x502)`

**Comment by gustopn** [comment]
Today a new AMDGPU-PRO driver was released 20.20 but it did not fix the issue.

**Comment by stefanodroghetti** [comment]
Yup, even the new Amdgpu-pro driver doesn't fix it.
It's a big problem, because without amdgpu-pro DaVinci Resolve doesn't work, and with amdgpu-pro OBS Studio doesn't work.
It's a bit hard to uninstall and reinstall amdgpu-pro each time one of those applications are needed.

P.S.: installing amdgpu-pro with --headless option fix the problem in OBS, and Blender or other software can be used with GPU compute. But still DaVinci resolve doesn't work.

**Comment by billyjbryant** [comment]
I am running into this bug as well, I upgraded graphics cards and in order to use the GPU for DaVinci, I need to have amdgpu-pro installed.  This is a crappy bug, but doesn't look like its OBS's fault.

**Comment by billyjbryant** [comment]
This is being tracked in some capacity here:  https://aur.archlinux.org/packages/opencl-amdgpu-pro-dev/

**Comment by stefanodroghetti** [comment]
New AMD drivers, new OBS version, still the same behaviour on November, 2020. Neither a workaround has been found. I'm seriously thinking about selling my AMD graphic card and buying an Nvidia :-(

**Comment by evilscientress** [comment]
@Stealthii the workaround is to not use the AMD Pro drivers. If you use the open source driver OBS works fine. I setup my system in a way that only applications like Davinci Resolve which need the AMD Pro drivers use the pro driver. All other applications use the open source driver. It's a bit hacky but by doing so everything works.

**Comment by stefanodroghetti** [comment]
> @Stealthii the workaround is to not use the AMD Pro drivers. If you use the open source driver OBS works fine. I setup my system in a way that only applications like Davinci Resolve which need the AMD Pro drivers use the pro driver. All other applications use the open source driver. It's a bit hacky but by doing so everything works.

Well this should be a great workaround. We do need to use DaVinci and OBS on a daily base. Can you share it?

**Comment by evilscientress** [comment]
@stefanodroghetti it's documented in the [Arch Linux wiki](https://wiki.archlinux.org/index.php/AMDGPU_PRO#Workaround_for_using_proprietary_OpenGL)

Essentially you have to install the proprietary OpenGL lib in a location where it's not loaded by default and run programs which need it like this:

```bash
$ env LD_LIBRARY_PATH="${PACKAGEPATH}/opt/amdgpu-pro/lib/x86_64-linux-gnu:${LD_LIBRARY_PATH}" LIBGL_DRIVERS_PATH="${PACKAGEPATH}/usr/lib/x86_64-linux-gnu/dri/" dri_driver="amdgpu" ${command}
```

I use DaVinci and OBS almost on a daily basis as well and I know your frustration. I was really glad when I found that workaround in the wiki. Hope it works for you as well.


I wanted to fix the texture issue in the obs xcomposite capture code myself, but sadly I don't know enough about opengl.

**Comment by stefanodroghetti** [comment]
Great! I was able to do it on Ubuntu 20.04. It's been much more complicated but it's ok.
Here's how I did it:

- Download the drivers from AMD site, unpack the file and enter the folder with the terminal.
- Uninstall the previous version:
`amdgpu-pro-uninstall`
- Install the drivers in the full version:
`./amdgpu-pro-install --opencl=legacy,pal`
- reboot
- Copy the two folders DaVinci needs to start:
```
sudo cp -R /opt/amdgpu-pro/lib/x86_64-linux-gnu /opt/amd1
sudo cp -R /usr/lib/x86_64-linux-gnu/dri /opt/amd2
```
- Uninstall AMD drivers:
`amdgpu-pro-uninstall`
- Install the drivers in headless mode:
`./amdgpu-pro-install --opencl=legacy,pal --headless`
- reboot

Now, every time I want to to run DaVinci I type:

`LD_LIBRARY_PATH="/opt/amd1" LIBGL_DRIVERS_PATH="/opt/amd2" dri_driver="amdgpu" /opt/resolve/bin/resolve`

Thank you everybody! I finally got it solved, so I don't have to sell my RX590 :-)
